SpatialDivisiveNormalization.lua

```diff
@@ -82,15 +82,33 @@ function SpatialDivisiveNormalization:__init(nInputPlane, kernel, threshold, thr
 end
 
 function SpatialDivisiveNormalization:updateOutput(input)
+   
+   self.localstds = self.stdestimator:updateOutput(input)
+
    -- compute side coefficients
-   if (input:size(3) ~= self.coef:size(3)) or (input:size(2) ~= self.coef:size(2)) then
-      local ones = input.new():resizeAs(input):fill(1)
-      self.coef = self.meanestimator:updateOutput(ones)
-      self.coef = self.coef:clone()
+   local dim = input:dim()
+   if self.localstds:dim() ~= self.coef:dim() or (input:size(dim) ~= self.coef:size(dim)) or (input:size(dim-1) ~= self.coef:size(dim-1)) then
+      self.ones = self.ones or input.new()
+      self.__coef = self.__coef or input.new()
+      if dim == 4 then
+         -- batch mode
+         self.ones:resizeAs(input[1]):fill(1)
+         local coef = self.meanestimator:updateOutput(self.ones)
+         self.__coef:resizeAs(coef):copy(coef) -- make contiguous
+         self._coef = self._coef or self.coef.new()
+         local coef = self.__coef:view(1,table.unpack(self.__coef:size():totable())):expandAs(self.localstds)
+         self._coef:resizeAs(coef):copy(coef)
```

Comment:
you dont need this resizeAs + copy anymore. the expandAs is good enough (I also should change it in batchnorm).
After the new cuda Apply kernels were landed, which support non-contiguous apply, expanded input which is non-contiguous is perfectly acceptable without introducing a sync point.